src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorLanguageQueryEndpoint.cs

```diff
@@ -0,0 +1,93 @@
+        {
+            return new TextDocumentIdentifier
+            {
+                Uri = request.Uri
+            };
+        }
+
+        public async Task<RazorLanguageQueryResponse> HandleRequestAsync(RazorLanguageQueryParams request, RazorRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            requestContext.RequireDocumentContext();
+            var documentContext = requestContext.DocumentContext;
+
+            var documentSnapshot = documentContext.Snapshot;
+            var documentVersion = documentContext.Version;
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync();
+            var sourceText = await documentSnapshot.GetTextAsync();
+            var linePosition = new LinePosition(request.Position.Line, request.Position.Character);
+            var hostDocumentIndex = sourceText.Lines.GetPosition(linePosition);
+            var responsePosition = request.Position;
+
+            if (codeDocument.IsUnsupported())
+            {
+                // All language queries on unsupported documents return Html. This is equivalent to what pre-VSCode Razor was capable of.
+                return new RazorLanguageQueryResponse()
+                {
+                    Kind = RazorLanguageKind.Html,
+                    Position = responsePosition,
+                    PositionIndex = hostDocumentIndex,
+                    HostDocumentVersion = documentVersion,
+                };
+            }
+
+            var responsePositionIndex = hostDocumentIndex;
+
+            var languageKind = _documentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex, rightAssociative: false);
+            if (languageKind == RazorLanguageKind.CSharp)
+            {
+                if (_documentMappingService.TryMapToProjectedDocumentPosition(codeDocument, hostDocumentIndex, out var projectedPosition, out var projectedIndex))
+                {
+                    // For C# locations, we attempt to return the corresponding position
+                    // within the projected document
+                    responsePosition = projectedPosition;
+                    responsePositionIndex = projectedIndex;
+                }
+                else
+                {
+                    // It no longer makes sense to think of this location as C#, since it doesn't
+                    // correspond to any position in the projected document. This should not happen
+                    // since there should be source mappings for all the C# spans.
+                    languageKind = RazorLanguageKind.Razor;
+                    responsePositionIndex = hostDocumentIndex;
+                }
+            }
+
+            requestContext.LspLogger.LogInformation("Language query request for ({requestPositionLine}, {requestPositionCharacter}) = {languageKind} at ({responsePositionLine}, {responsePositionCharacter})",
```

Comment:
It's so weird reviewing code that doesn't compile. I feel like i'm taking crazy pills 🤣
```suggestion
            requestContext.Logger.LogInformation("Language query request for ({requestPositionLine}, {requestPositionCharacter}) = {languageKind} at ({responsePositionLine}, {responsePositionCharacter})",
```

Reply:
Agreed about the non-compiling, but actually in this case both would compile. I'll be conjoining LspLogger and Logger eventually.